torchmetrics/image/map.py

```diff
@@ -0,0 +1,369 @@
+                    [num_boxes, 4] containing `num_boxes` groundtruth boxes of the format
+                    [xmin, ymin, xmax, ymax] in absolute image coordinates.
+                - ``labels``: torch.IntTensor of shape
+                    [num_boxes] containing 1-indexed groundtruth classes for the boxes.
+
+        Raises:
+            ValueError:
+                If ``preds`` is not of type List[Dict[str, torch.Tensor]]
+            ValueError:
+                If ``target`` is not of type List[Dict[str, torch.Tensor]]
+            ValueError:
+                If `preds` and `target` are not of the same length
+            ValueError:
+                If any of `preds.boxes`, `preds.scores`
+                and `preds.labels` are not of the same length
+            ValueError:
+                If any of `target.boxes` and `target.labels` are not of the same length
+            ValueError:
+                If any box is not type float and of length 4
+            ValueError:
+                If any class is not type int and of length 1
+            ValueError:
+                If any score is not type float and of length 1
+        """
+        _input_validator(preds, target)
+
+        for item in preds:
+            self.detection_boxes.append(item["boxes"])
+            self.detection_scores.append(item["scores"])
+            self.detection_labels.append(item["labels"])
+
+        for item in target:
+            self.groundtruth_boxes.append(item["boxes"])
+            self.groundtruth_labels.append(item["labels"])
+
+    def compute(self) -> MAPMetricResults:
+        """Compute the `Mean-Average-Precision (mAP) and Mean-Average-Recall (mAR)` scores. All detections added in
+        the `update()` method are included.
+
+        .. note::
+            Scores are calculated with @[ IoU=0.50:0.95 | area=all | maxDets=100 ]
+
+        Returns:
+            dict containing
+
+            - map_value: ``torch.Tensor``
+            - mar_value: ``torch.Tensor``
+            - map_per_class_value: ``List[torch.Tensor]``
+            - mar_per_class_value: ``List[torch.Tensor]``
+        """
+        coco_target, coco_preds = COCO(), COCO()
+        coco_target.dataset = self._get_coco_format()
+        coco_preds.dataset = self._get_coco_format(is_pred=True)
+
+        with _hide_prints():
+            coco_target.createIndex()
+            coco_preds.createIndex()
+            coco_eval = COCOeval(coco_target, coco_preds, "bbox")
+            coco_eval.evaluate()
+            coco_eval.accumulate()
+            coco_eval.summarize()
+            stats = coco_eval.stats
+
+        new_average_precision = torch.cat(
+            (
+                self.average_precision,
+                torch.tensor(
+                    [stats[COCO_STATS_MAP_VALUE_INDEX]], dtype=torch.float, device=self.average_precision.device
+                ),
+            )
+        )
+        setattr(self, "average_precision", new_average_precision)
```

Comment:
Why not ``self.average_precision = new_average_precision``?

Reply:
I had to use the setattr for the `class_map_values` so I thought it would be nicer to have this aligned.

No personal preference. Do you prefer `self.average_precision = new_average_precision` ?

Reply:
Resolved as of https://github.com/PyTorchLightning/metrics/pull/467#discussion_r714596707